dask/dataframe/io/csv.py

```diff
@@ -68,6 +68,8 @@ def project_columns(self, columns):
         """Return a new CSVFunctionWrapper object with
         a sub-column projection.
         """
+        # Make sure columns is ordered correctly
+        columns = [c for c in self.head.columns if c in columns]
```

Comment:
This could get a little tricky if there are duplicated names either in the CSV or the projected columns list. In both cases, I think you could wind up with a list that is not the same length as the original.

I haven't fully thought through the failure modes, as it's getting kind of late, so treat this more as a placeholder thought :)

Reply:
Yes- We certainly assume that column names are unique in a pandas DataFrame (which I believe is a requirement), and the `columns` argument passed in the `project_columns` should always be a unique list.

Reply:
> Yes- We certainly assume that column names are unique in a pandas DataFrame (which I believe is a requirement)

I don't think it's a requirement of dataframes, I can happily create one by inserting a new column into your test df, or with
```python
import pandas as pd
df1 = pd.DataFrame({"a": [1]})
df2 = pd.concat((df1, df1), axis=1)
```

It seems that there is some history to what `pd.read_csv` does about this re: `mangle_dupe_cols` (currently it renames duplicates by default).

> and that the `columns` argument passed in to `project_columns` is always be a unique list.

Do we have to assume this? I can select columns more than once out of a dataframe

Reply:
>I can happily create one by inserting a new column into your test df, or with

Ah - you are right about that. I was thinking of cudf (which doesn't support duplicate columns, since supporting such a thing is more problematic than useful).

>Do we have to assume this? I can select columns more than once out of a dataframe

We are not actually performing the final column selection in `project_columns`, just avoiding unnecessary IO. The follow-up getitem operation performs the actual column selection.

Overall, there is no need to support duplicate column selection in `project_columns`, but it seems we do need to support the case that the original dataframe has duplicate columns...

Reply:
As far as I can tell, pandas doesn't actually support [mangle_dupe_cols=False](https://github.com/pandas-dev/pandas/blob/2cb96529396d93b46abab7bbc73a208e708c642e/pandas/io/parsers.py#L833).  Therefore, I don't think there are any cases where a direct `read_csv` call will produce duplicate columns.